python/ray/serve/autoscaling_policy.py

```diff
@@ -142,136 +107,59 @@ class ReplicaQueueLengthAutoscalingPolicy(AutoscalingPolicy):
-            if self.decision_counter < -self.scale_down_consecutive_periods:
-                self.decision_counter = 0
-                decision_num_replicas = desired_num_replicas
-
-        # Do nothing.
-        else:
-            self.decision_counter = 0
-
-        return decision_num_replicas
-
-    def apply_bounds(
-        self,
-        curr_target_num_replicas: int,
-        target_capacity: Optional[float] = None,
-        target_capacity_direction: Optional[TargetCapacityDirection] = None,
-    ) -> int:
-        """Clips curr_target_num_replicas using the current bounds."""
-
-        upper_bound = get_capacity_adjusted_num_replicas(
-            self.config.max_replicas,
-            target_capacity,
-        )
-        lower_bound = self.get_current_lower_bound(
-            target_capacity, target_capacity_direction
-        )
-        return max(lower_bound, min(upper_bound, curr_target_num_replicas))
-
-    def get_current_lower_bound(
-        self,
-        target_capacity: Optional[float] = None,
-        target_capacity_direction: Optional[TargetCapacityDirection] = None,
-    ) -> int:
-        """Get the autoscaling lower bound, including target_capacity changes.
+    decision_num_replicas = curr_target_num_replicas
 
-        The autoscaler uses initial_replicas scaled by target_capacity only
-        if the target capacity direction is UP.
-        """
+    desired_num_replicas = _calculate_desired_num_replicas(
+        config,
+        current_num_ongoing_requests,
+        override_min_replicas=capacity_adjusted_min_replicas,
+        override_max_replicas=capacity_adjusted_max_replicas,
+    )
+    # Scale up.
+    if desired_num_replicas > curr_target_num_replicas:
+        # If the previous decision was to scale down (the counter was
+        # negative), we reset it and then increment it (set to 1).
+        # Otherwise, just increment.
+        if decision_counter < 0:
+            decision_counter = 0
+        decision_counter += 1
+
+        # Only actually scale the replicas if we've made this decision for
+        # 'scale_up_consecutive_periods' in a row.
+        if decision_counter > int(config.upscale_delay_s / CONTROL_LOOP_PERIOD_S):
+            decision_counter = 0
+            decision_num_replicas = desired_num_replicas
+
+    # Scale down.
+    elif desired_num_replicas < curr_target_num_replicas:
+        # If the previous decision was to scale up (the counter was
+        # positive), reset it to zero before decrementing.
+        if decision_counter > 0:
+            decision_counter = 0
+        decision_counter -= 1
+
+        # Only actually scale the replicas if we've made this decision for
+        # 'scale_down_consecutive_periods' in a row.
+        if decision_counter < -int(config.downscale_delay_s / CONTROL_LOOP_PERIOD_S):
+            decision_counter = 0
+            decision_num_replicas = desired_num_replicas
+
+    # Do nothing.
+    else:
+        decision_counter = 0
```

Comment:
This should be deferred to PolicyManager to whether accept, reject or delay scaling recommendation produced by the policy. Policy by itself should be stateless, though could be accepting historical data (say last 5min of rersource usage) to take its decision.

Reply:
Totally agreed the policy itself should be stateless and that's why it's been refactored to just a function. However, I don't think the policy manager should decide how fast or slow it the replica scales. There might be usecase that the user is tracking some sort of session count on the main page and wanting to warm up/ cool down the ML service by scaling the deployment replica up/ down accordingly. I don't think we should stop customer from quickly scale up/ down by moving this logics out.

Reply:
Yes, policy should be able to control how fast we auto-scale, but policy shouldn't be relying on the frequency of its invocation (`CONTROL_LOOP_PERIOD_S`)

Reply:
Agreed, this is currently out of scope of the custom autoscaling project, but we can follow up to rewrite this logics to use the exact time when autoscale last happened :)

Reply:
We can persist history w/in policy_state for ex
https://github.com/ray-project/ray/pull/42284#discussion_r1449405138